contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java

```diff
@@ -0,0 +1,235 @@
+package org.apache.drill.exec.store.http.util;
+
+import okhttp3.Cache;
+import okhttp3.Credentials;
+import okhttp3.FormBody;
+import okhttp3.Interceptor;
+import okhttp3.OkHttpClient;
+import okhttp3.OkHttpClient.Builder;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.exec.ExecConstants;
+import org.apache.drill.exec.ops.FragmentContext;
+import org.apache.drill.exec.store.http.HttpAPIConfig;
+import org.apache.drill.exec.store.http.HttpStoragePluginConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+
+/**
+ * This class performs the actual HTTP requests for the HTTP Storage Plugin. The core method is the getInputStream()
+ * method which accepts a url and opens an InputStream with that URL's contents.
+ */
+public class SimpleHttp {
+  private static final Logger logger = LoggerFactory.getLogger(SimpleHttp.class);
+
+  private final OkHttpClient client;
+
+  private final HttpStoragePluginConfig config;
+
+  private final FragmentContext context;
+
+  private final HttpAPIConfig apiConfig;
+
+  public SimpleHttp(HttpStoragePluginConfig config, FragmentContext context, String connectionName) {
+    this.config = config;
+    this.context = context;
+    this.apiConfig = config.connections().get(connectionName);
+    client = setupServer();
+  }
+
+
+
+  public InputStream getInputStream(String urlStr) {
+    Request.Builder requestBuilder;
+
+    // The configuration does not allow for any other request types other than POST and GET.
+    if (apiConfig.method().equals("get")) {
+      // Handle GET requests
+      requestBuilder = new Request.Builder().url(urlStr);
+    } else {
+      // Handle POST requests
+      FormBody.Builder formBodyBuilder = buildPostBody();
+      requestBuilder = new Request.Builder()
+        .url(urlStr)
+        .post(formBodyBuilder.build());
+    }
+
+    // Add headers to request
+    if (apiConfig.headers() != null) {
+      for (Map.Entry<String, String> entry : apiConfig.headers().entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+        requestBuilder.addHeader(key, value);
+      }
+    }
+
+    // Build the request object
+    Request request = requestBuilder.build();
+    logger.debug("Headers: {}", request.headers());
+
+    try {
+      // Execute the request
+      Response response = client
+        .newCall(request)
+        .execute();
+      logger.debug(response.toString());
+
+      // If the request is unsuccessful, throw a UserException
+      if (!response.isSuccessful()) {
+        throw UserException.dataReadError()
+          .message("Error retrieving data: " + response.code() + " " + response.message())
+          .addContext("Response code: ", response.code())

```

Comment:
Please add more context. Which plugin? Which table? Which config? Else, all the user knows is "something went wrong". A person helping would not know which plugin, etc. without asking the user for the query, then working out manually these important details.

The EVF has a way of handling this kind of context generically. Turns out you can use that same mechanism here if useful, since most of it relies on features of `UserException`.

Reply:
I added more info here.  Unfortunately the response isn't available below, but I'd think there's enough info to debug.  At a minimum the user will see that it was the HTTP plugin and the url string that is being sent.  They also may get the response code and message.

Reply:
Is there something more specific you're looking for?